### ISSUE TITLE
Large numbers in heterogeneous datasets can raise errors on decompression

### ISSUE THREAD

**Issue by clbarnes** [constantinpape/z5#58]
This is a really bizarre bug. Zlib throws errors intermittently on decompression, but

1. Only when there are large numbers in the dataset
2. Not if the dataset is full of the same number

I'm testing on uint8 data and the threshold for what's a large number seems to be 181.

Error message is 

```
Traceback (most recent call last):
  File "failing_case.py", line 7, in <module>
    print(ds[:])
  File "/home/barnesc/.pyenv/versions/heinrich3/lib/python3.6/site-packages/z5py/dataset.py", line 366, in __getitem__
    n_threads=self.n_threads)
RuntimeError: Exception during zlib decompression: (-5) 
```

Here's an example sweep I did with `np.ones * (max_ - 1)` as the homogeneous data, and random integers between 0 and `max_` (keep generating until you get at least one of `max_ - 1` in there).

5 reps each, output to be read like a unit test. This pattern is pretty consistent.

```
{
  178:{
    'homogeneous':'.....',
    'heterogeneous':'.....'
  },
  179:{
    'homogeneous':'.....',
    'heterogeneous':'.....'
  },
  180:{
    'homogeneous':'.....',
    'heterogeneous':'...e.'
  },
  181:{
    'homogeneous':'.....',
    'heterogeneous':'ee..e'
  },
  182:{
    'homogeneous':'.....',
    'heterogeneous':'eeeee'
  },
  183:{
    'homogeneous':'.....',
    'heterogeneous':'eeeee'
  },
  184:{
    'homogeneous':'.....',
    'heterogeneous':'eeeee'
  }
}
```

Script here https://gist.github.com/clbarnes/d51db4731d32c23401afa70be5981e7a

**Comment by clbarnes** [comment]
I haven't experimented with different data types but could give it a go.

Reading real data from the L1 export doesn't seem to be an issue, so I should also try to see whether blcoks written with z5 can be read by other libraries (like rust-n5).

**Comment by constantinpape** [comment]
Yes, seeing if these issues occur for different datatypes would help.

I have tests that check that data can be read with java-n5:
https://github.com/constantinpape/z5/blob/master/src/test/test_n5/test_n5.cxx

These are not included in the travis unit tests though, because I didn't want to include all the extra dependencies in travis.

**Comment by clbarnes** [comment]
Floats do not seem to be affected; both signed and unsigned integers do. Larger integer types are affected the same as the uint8 example. The threshold for larger integer types is larger in an absolute sense but occurs lower in the range of that dtype's values.

The really bizarre thing is that it seems to be dependent on the arrangement of numbers in the array. I'm now generating heterogeneous data by populating a 1D array of the right length and then reshaping it. If it's in sorted order before the reshape, I get no errors. If it's shuffled, I get the errors.

**Comment by constantinpape** [comment]
Thanks for checking for larger integers.
The error code that is returned (-5) means Z_BUF_ERR, which translates, for decompression, to not enough space available in the output buffer.

My best guess is that the heterogenous data you produce can only be compressed by a 
small degree, which makes some (potentially intermediate) buffer overflow.
This would explain why it does not occur for homogenous data or why shuffling makes a difference.
I will try to have a closer look at this tomorrow or on the weekend.

Regarding `n_threads > 1`, does this appear independently from the compression bug?
I.e. do you get the segfault every time you try to use multi-threaded I/O?
If so, the library probably does not link properly against libthread.

**Comment by constantinpape** [comment]
This issue was caused by a too small output buffer size.
I increased it fron 32KB to 256KB (which is also recommended in the zlib user guide):
https://github.com/constantinpape/z5/blob/release1.0/include/z5/compression/zlib_compressor.hxx#L40

This fixed the issue; at least for uint8. I haven't tried for larger integers yet.
I will run test for the other integers on the weekend, gonna leave it open until then.
